app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt

```diff
@@ -39,33 +47,77 @@ abstract class BaseGitActivity : AppCompatActivity() {
      * @param operation The type of git operation to launch
      */
-    suspend fun launchGitOperation(operation: Int) {
+    suspend fun launchGitOperation(operation: Int): Result<Unit> {
         if (GitSettings.url == null) {
-            setResult(RESULT_CANCELED)
-            finish()
-            return
+            return Result.failure(IllegalStateException("Git url is not set!"))
         }
-        try {
-            val localDir = requireNotNull(PasswordRepository.getRepositoryDirectory())
-            val op = when (operation) {
-                REQUEST_CLONE, GitOperation.GET_SSH_KEY_FROM_CLONE -> CloneOperation(localDir, GitSettings.url!!, this)
-                REQUEST_PULL -> PullOperation(localDir, this)
-                REQUEST_PUSH -> PushOperation(localDir, this)
-                REQUEST_SYNC -> SyncOperation(localDir, this)
-                BREAK_OUT_OF_DETACHED -> BreakOutOfDetached(localDir, this)
-                REQUEST_RESET -> ResetToRemoteOperation(localDir, this)
-                else -> {
-                    tag(TAG).e { "Operation not recognized : $operation" }
-                    setResult(RESULT_CANCELED)
-                    finish()
-                    return
-                }
+        val op = when (operation) {
+            REQUEST_CLONE, GitOperation.GET_SSH_KEY_FROM_CLONE -> CloneOperation(this, GitSettings.url!!)
+            REQUEST_PULL -> PullOperation(this)
+            REQUEST_PUSH -> PushOperation(this)
+            REQUEST_SYNC -> SyncOperation(this)
+            BREAK_OUT_OF_DETACHED -> BreakOutOfDetached(this)
+            REQUEST_RESET -> ResetToRemoteOperation(this)
+            else -> {
+                tag(TAG).e { "Operation not recognized : $operation" }
+                return Result.failure(IllegalArgumentException("$operation is not a valid Git operation"))
             }
-            op.executeAfterAuthentication(GitSettings.authMode)
-        } catch (e: Exception) {
-            e.printStackTrace()
-            MaterialAlertDialogBuilder(this).setMessage(e.message).show()
         }
+        return op.executeAfterAuthentication(GitSettings.authMode)
+    }
+
+    fun defaultSuccessHandler(@Suppress("UNUSED_PARAMETER") nothing: Unit) {
+        finish()
+    }
+
+    fun defaultErrorHandler(err: Throwable) {
+        val error = rootCauseException(err)
+        if (!isExplicitlyUserInitiatedError(error)) {
+            getEncryptedPrefs("git_operation").edit {
+                remove(PreferenceKeys.HTTPS_PASSWORD)
+            }
+            sharedPrefs.edit { remove(PreferenceKeys.SSH_OPENKEYSTORE_KEYID) }
+            d(error)
+            MaterialAlertDialogBuilder(this)
+                .setTitle(resources.getString(R.string.jgit_error_dialog_title))
+                .setMessage(ErrorMessages[error])
+                .setPositiveButton(resources.getString(R.string.dialog_ok)) { _, _ ->
+                    finish()
+                }.show()
+        }
+    }
```

Comment:
I added these to simplify general cases for all callers to mirror the now defunct `onSuccess`/`onError` API from `GitOperation`. It unfortunately now needs to be called manually by each caller but I don't have any better ideas.

Reply:
I also don't know how to improve that design and think it's okay now that it happens on the caller side.

How about we name these `finishOnSuccessHandler` and `finishAfterPromptOnErrorHandler` (or something similar) though?

Reply:
Sure I'll rename these in a bit.